### ISSUE TITLE
semantics of send/receive

### ISSUE THREAD

**Issue by jacobmerson** [SCOREC/redev#18]
@cwsmith 
What are the semantics of adios send/recieve? Are those operations blocking in general? I.e. in RDV usage is communicator.send() blocking? If not, how do you check if the send has completed?

**Comment by cwsmith** [comment]
`communicator.Send()` and `communicator.Recv()` are blocking given that they call ADIOS2 `Perform[Puts|Gets]` and `EndStep`.

https://github.com/SCOREC/redev/blob/65f42def5c48343532bc998ed934b195a99c2db9/redev_comm.h#L233-L234
https://github.com/SCOREC/redev/blob/65f42def5c48343532bc998ed934b195a99c2db9/redev_comm.h#L280-L281

ADIOS2 `Put` and `Get` are not blocking when using the SST transport and the default [`Deferred` mode](https://adios2.readthedocs.io/en/latest/components/components.html#put-modes-and-memory-contracts).  The call to `[EndStep()](https://adios2.readthedocs.io/en/latest/components/components.html#endstep)`, `PerformPuts()`, `PerformGets()` are blocking and returns when the data transfer is complete.

More details on `SST` are here:
https://adios2.readthedocs.io/en/latest/engines/engines.html#sst-sustainable-staging-transport

**Comment by cwsmith** [comment]
Yeah, the abandoned `Pack(...)` API would have accommodated this.  

I *think* we can split the `Put` and `PerformPuts/EndStep` calls.  I'll take a closer look.

**Comment by jacobmerson** [comment]
I guess the next question is does this make sense for the use case we expect where you have many engines that are used...one for each field. Vs. many variables in a single engine.

**Comment by cwsmith** [comment]
Right.  Good point.  If each field has an Engine/Comm then we don't need to split the `Put` `PerformPuts` calls.
